overlord/hookstate/ctlcmd/fail_test.go

```diff
@@ -0,0 +1,127 @@
+
+	"gopkg.in/check.v1"
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/features"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/hookstate/ctlcmd"
+	"github.com/snapcore/snapd/overlord/registrystate"
+	"github.com/snapcore/snapd/snap"
+)
+
+func (s *registrySuite) mockRegistriesFeature(c *C) (restore func()) {
+	oldDir := dirs.FeaturesDir
+	dirs.FeaturesDir = c.MkDir()
+
+	registryCtlFile := features.Registries.ControlFile()
+	c.Assert(os.WriteFile(registryCtlFile, []byte(nil), 0644), check.IsNil)
+
+	return func() {
+		c.Assert(os.Remove(registryCtlFile), IsNil)
+		dirs.FeaturesDir = oldDir
+	}
+}
+
+func (s *registrySuite) TestFailAbortsRegistryTransaction(c *C) {
+	restore := s.mockRegistriesFeature(c)
+	defer restore()
+
+	s.state.Lock()
+	chg := s.state.NewChange("test", "")
+	commitTask := s.state.NewTask("commit-registry-tx", "")
+	chg.AddTask(commitTask)
+	tx, err := registrystate.NewTransaction(s.state, "my-acc", "my-reg")
+	c.Assert(err, IsNil)
+
+	err = tx.Set("foo", "bar")
+	c.Assert(err, IsNil)
+
+	commitTask.Set("registry-transaction", tx)
+
+	task := s.state.NewTask("run-hook", "")
+	chg.AddTask(task)
+	setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "change-view-plug"}
+	task.Set("commit-task", commitTask.ID())
+	s.state.Unlock()
+
+	mockContext, err := hookstate.NewContext(task, s.state, setup, s.mockHandler, "")
+	c.Assert(err, IsNil)
+
+	stdout, stderr, err := ctlcmd.Run(mockContext, []string{"fail", "don't like changes"}, 0)
+	c.Assert(err, IsNil)
+	c.Check(stdout, IsNil)
+	c.Check(stderr, IsNil)
+
+	tx = nil
+	s.state.Lock()
+	err = commitTask.Get("registry-transaction", &tx)
+	s.state.Unlock()
+	c.Assert(err, IsNil)
+
+	snap, reason := tx.AbortInfo()
+	c.Assert(reason, Equals, "don't like changes")
+	c.Assert(snap, Equals, "test-snap")
+}
+
+func (s *registrySuite) TestFailErrors(c *C) {
+	stdout, stderr, err := ctlcmd.Run(s.mockContext, []string{"fail", "reason"}, 0)
+	c.Assert(err, ErrorMatches, i18n.G(`cannot use "snapctl fail" without enabling the "experimental.registries" feature`))
+	c.Check(stdout, IsNil)
+	c.Check(stderr, IsNil)
+
+	restore := s.mockRegistriesFeature(c)
+	defer restore()
+
+	stdout, stderr, err = ctlcmd.Run(s.mockContext, []string{"fail"}, 0)
+	c.Assert(err, ErrorMatches, i18n.G("the required argument `:<rejection-reason>` was not provided"))
+	c.Check(stdout, IsNil)
+	c.Check(stderr, IsNil)
+
+	s.state.Lock()
+	task := s.state.NewTask("run-hook", "my test task")
+	setup := &hookstate.HookSetup{Snap: "test-snap", Revision: snap.R(1), Hook: "other-hook"}
+	s.state.Unlock()
+
+	s.mockContext, err = hookstate.NewContext(task, s.state, setup, s.mockHandler, "")
+	c.Assert(err, IsNil)
+
+	stdout, stderr, err = ctlcmd.Run(s.mockContext, []string{"fail", "reason"}, 0)
+	c.Assert(err, ErrorMatches, i18n.G(`cannot use "snapctl fail" outside of a "change-view" hook`))
```

Comment:
maybe we could also test for the ephemeral case?

Reply:
Added, thanks